### PR TITLE
Customised logger with io.writer

### LINKED ISSUE #2892: Refine logging in integration tests, use customized logger with GinkgoWriter
The library `controller-runtime` requires setting a logger (by `log.SetLogger()`) at the first 30s when the application starts, or it would use the default `NullLogSink`. We should also call it in testing codes.

When we test with ginkgo, ginkgo provides a helpful `GinkgoWriter`, which hides the output as default, only prints it when the test failed. We'd better use it to keep our testing output clean.

So here is the proposal:

- Create another constructor in `pkg/log/zap.go` like `NewDefaultZapLogger` but with the custom `io.Writer`.
- replace `log.SetLogger()` with above logger

Here are several files that need changes:

- api/v1alpha1/suite_test.go
- cmd/chaos-controller-manager/provider/suit_test.go
- controllers/common/suit_test.go
- controllers/schedule/suit_test.go
- pkg/webhook/inject/setup_test.go
- pkg/workflow/controllers/suite_test.go

About using customized with zap, you could create a logger by using `zapcore` directly. Or using `zap.Config` with registering custom sink factory with [RegisterSink](https://pkg.go.dev/go.uber.org/zap#RegisterSink). Reference: https://github.com/uber-go/zap/issues/979

### PR DESCRIPTION
**Pull request by basit9958**
### What problem does this PR solve?

# fixes #2892

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] Breaking backward compatibility

### CODE REVIEW

File: pkg/log/zap.go [hunk 1]
```diff
@@ -32,3 +37,21 @@ func NewDefaultZapLogger() (logr.Logger, error) {
 	logger := zapr.NewLogger(zapLogger)
 	return logger, nil
 }
+
+func NewZapLoggerWithWriter() logr.Logger {
```
Comment (STRRL) [inline pkg/log/zap.go:41]: please move the `writer` as the parameter of this constructor

Comment (STRRL) [reply pkg/log/zap.go:41]: and we would use `ginkgo.GinkgoWriter` within integration test codes

**Review by STRRL — CHANGES_REQUESTED** [review]

**Comment by STRRL** [comment]
Hi @basit9958, do you need some help?

File: api/v1alpha1/suite_test.go [hunk 2]
```diff
@@ -48,9 +48,9 @@ func TestAPIs(t *testing.T) {
 
 var _ = BeforeSuite(func(done Done) {
 
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
-
-	By("bootstrapping test environment")
+	//logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
```
Comment (STRRL) [inline api/v1alpha1/suite_test.go:51]: please remove this line of comment

File: cmd/chaos-controller-manager/provider/suit_test.go [hunk 3]
```diff
@@ -58,7 +58,8 @@ func TestProvider(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	//logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
```
Comment (STRRL) [inline cmd/chaos-controller-manager/provider/suit_test.go:61]: please remove this line of comment

File: controllers/common/suit_test.go [hunk 4]
```diff
@@ -66,7 +66,8 @@ func TestCommon(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	//logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
```
Comment (STRRL) [inline controllers/common/suit_test.go:69]: please remove this line of comment

File: api/v1alpha1/suite_test.go [hunk 5]
```diff
@@ -48,9 +48,9 @@ func TestAPIs(t *testing.T) {
...
+	logf.setLogger(log.NewZapLoggerWithWriter(ginkgo.GinkgoWriter))
+		By("bootstrapping test environment")
```
Comment (STRRL) [inline api/v1alpha1/suite_test.go:53]: please fix indent level of this line of code

File: pkg/webhook/inject/setup_test.go [hunk 6]
```diff
@@ -52,8 +51,9 @@ func TestAPIs(t *testing.T) {
 }
 
 var _ = BeforeSuite(func(done Done) {
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 
+	//logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
```
Comment (STRRL) [inline pkg/webhook/inject/setup_test.go:55]: please remove this line of comment

File: pkg/workflow/controllers/suite_test.go [hunk 7]
```diff
@@ -56,7 +56,8 @@ func TestWorkflow(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	//logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
```
Comment (STRRL) [inline pkg/workflow/controllers/suite_test.go:59]: please remove this line of comment

File: pkg/log/zap.go [hunk 8]
```diff
@@ -32,3 +34,18 @@ func NewDefaultZapLogger() (logr.Logger, error) {
...
+func NewZapLoggerWithWriter(out io.Writer) logr.Logger {
+
+	bWriter := out
+	config := zap.NewDevelopmentConfig()
+	config.EncoderConfig.EncodeLevel = zapcore.CapitalLevelEncoder
+	config.EncoderConfig.FunctionKey = "function"
+	
+	core := zapcore.NewCore(zapcore.NewJSONEncoder(config.EncoderConfig), zapcore.AddSync(bWriter), config.Level)
+	zapLogger := zap.New(core)
+	zapLogger.Error("an error")
```
Comment (STRRL) [inline pkg/log/zap.go:47]: please remove this line of code

**Review by STRRL — CHANGES_REQUESTED** [review]
`logf.setLogger(log.NewZapLoggerWithWriter(ginkgo.GinkgoWriter))` using unexported method `setLogger`, it should be `SetLogger`

please fix the related imports issue, some package already imported as `.`, and `pkg/log` might need an alias

File: pkg/log/zap.go (hunk 8 above)
Comment (basit9958) [reply pkg/log/zap.go:47]: @STRRL  Can you please elaborate on how should i create a logger if i removed these three lines ?
or i can use flags to create a logger like this 
`typeNewZapLoggerWithWriter {
	out  log.Logger
}
func NewZapLoggerWithWriter(w io.Writer) logr.Logger{
	return &simpleLogger{
		out: log.New(w, "", log.LstdFlags|log.Lmicroseconds),
	}
}


`

Comment (STRRL) [reply pkg/log/zap.go:47]: please only remove L47 :)

**Comment by STRRL** [comment]
Hi @basit9958, do you need some help?

**Review by STRRL — APPROVED** [review]
LGTM!

**Comment by basit9958** [comment]
Yay!

**Comment by STRRL** [comment]
> Yay!

Oops! There are some compiler errors reported by CI. Please fix that. I am sorry for missing hat. 🤣

**Comment by STRRL** [comment]
you could run `make check` on you local machine, and fix the reported error.

**Comment by basit9958** [comment]
> you could run `make check` on your local machine, and fix the reported error.

Okay 👌

File: cmd/chaos-controller-manager/provider/suit_test.go [hunk 9]
```diff
@@ -58,7 +58,7 @@ func TestProvider(t *testing.T) {
 }
 
 var _ = BeforeSuite(func() {
-	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+	logf.SetLogger(log.NewZapLoggerWithWriter(ginkgo.GinkgoWriter))
```
Comment (STRRL) [inline cmd/chaos-controller-manager/provider/suit_test.go:61]:
```suggestion
	logf.SetLogger(log.NewZapLoggerWithWriter(GinkgoWriter))
```

File: pkg/webhook/inject/setup_test.go [hunk 10]
```diff
@@ -21,6 +21,8 @@ import (
 	"testing"
 	"time"
 
+	"github.com/chaos-mesh/chaos-mesh/pkg/log"
```
Comment (STRRL) [inline pkg/webhook/inject/setup_test.go:24]: we need an alias for this import because the symbol `log` is already used under this package.

**Review by STRRL — CHANGES_REQUESTED** [review]

File: cmd/chaos-controller-manager/provider/suit_test.go (hunk 9 above)
Comment (basit9958) [reply cmd/chaos-controller-manager/provider/suit_test.go:61]: This should be done only in this file or every file @STRRL ?

Comment (STRRL) [reply cmd/chaos-controller-manager/provider/suit_test.go:61]: for every file which uses dot imports with "github.com/onsi/ginkgo"

**Pull request closed without merging**